auth-lib/src/main/java/com/spotify/sdk/android/auth/TokenExchangeRequest.java

```diff
@@ -0,0 +1,221 @@
+        mCode = code;
+        mRedirectUri = redirectUri;
+        mCodeVerifier = codeVerifier;
+    }
+
+    /**
+     * Executes the token exchange request synchronously.
+     * This method performs a blocking HTTP request and should not be called on the main thread.
+     *
+     * @return TokenExchangeResponse containing the access token or error information
+     */
+    @NonNull
+    public TokenExchangeResponse execute() {
+        HttpURLConnection connection = null;
+        try {
+            final URL url = new URL(TOKEN_ENDPOINT);
+            connection = (HttpURLConnection) url.openConnection();
+            
+            connection.setRequestMethod("POST");
+            connection.setRequestProperty("Content-Type", CONTENT_TYPE_FORM);
+            connection.setDoOutput(true);
+            connection.setConnectTimeout(TIMEOUT_MS);
+            connection.setReadTimeout(TIMEOUT_MS);
+
+            final String requestBody = buildRequestBody();
+            
+            try (final OutputStream outputStream = connection.getOutputStream()) {
+                outputStream.write(requestBody.getBytes(StandardCharsets.UTF_8));
+                outputStream.flush();
+            }
+
+            final int responseCode = connection.getResponseCode();
+            final String responseBody = readResponse(connection, responseCode >= 400);
+            
+            return TokenExchangeResponse.fromHttpResponse(responseCode, responseBody);
+
+        } catch (final IOException e) {
+            return TokenExchangeResponse.fromError("network_error", "Network error: " + e.getMessage());
+        } finally {
+            if (connection != null) {
+                connection.disconnect();
+            }
+        }
+    }
+
+    @NonNull
+    private String buildRequestBody() {
+        try {
+            return "grant_type=" + URLEncoder.encode(GRANT_TYPE_AUTHORIZATION_CODE, "UTF-8") +
+                   "&client_id=" + URLEncoder.encode(mClientId, "UTF-8") +
+                   "&code=" + URLEncoder.encode(mCode, "UTF-8") +
+                   "&redirect_uri=" + URLEncoder.encode(mRedirectUri, "UTF-8") +
+                   "&code_verifier=" + URLEncoder.encode(mCodeVerifier, "UTF-8");
+        } catch (final Exception e) {
+            // This should never happen with UTF-8
+            throw new RuntimeException("Failed to encode request parameters", e);
+        }
+    }
+
+    @NonNull
+    private String readResponse(@NonNull final HttpURLConnection connection, final boolean isError) throws IOException {
+        final BufferedReader reader = new BufferedReader(new InputStreamReader(
+                isError ? connection.getErrorStream() : connection.getInputStream(),
+                StandardCharsets.UTF_8));
+        
+        final StringBuilder response = new StringBuilder();
+        String line;
+        while ((line = reader.readLine()) != null) {
+            response.append(line);
+        }
+        reader.close();
```

Comment:
We should probably wrap that around a try/catch/finally or use try-with-resources to ensure the InputStream isn't leaked when an exception is thrown.